### ISSUE TITLE
Proposal: minimal/embedded/lean Sentry for integration tests

### ISSUE THREAD

**Issue by kokes** [getsentry/sentry#32794]
### Problem Statement

We want to test our internal logic that collects various bits and bobs before sending it to Sentry. We can (and do) mock the server to verify the payload in our unit tests, but we'd also like to send our payload all the way to a test project in Sentry, to check Sentry will accept it.

While we can send it directly to the SaaS endpoint, we'd prefer to run these tests against a local instance. Since this is pretty much stateless (we only need a working API endpoint), the service can be barebones. Unfortunately, the recommended [devservices](https://develop.sentry.dev/testing/) is anything but. The inclusion of several databases, caches, message queues etc. makes this a pretty heavyweight solution.

Ideally the workflow would be like this (in pseudo Dockerfile syntax)

```
FROM python:3.x
RUN python -m pip install sentry[local]
EXPOSE ...
CMD sentry --local run xyz
```

(Note the lack of xmlsec installation, more on that in the proposed solution below.)

### Solution Brainstorm

There are several steps to this solution (in no particular order):

1. Abstract away all the database interaction (like Arroyo does already) and wrap all db driver imports
2. Provide local alternatives - be it in-memory data structures (for Redis or memcached) or in-memory/on-disk databases like SQLite (for Postgres, possibly Clickhouse, if needed)
3. Create setuptools extras to install all the dependencies like now, _apart from_ all the database drivers, docker lib and python3-saml - the last one to avoid needing a system-level dependency (xmlsec)
4. Migrate devservices to this dockerless workflow (or at least offer it as an option)

I tried to take a stab at it, but not being super familiar with Sentry's architecture, I hit dead ends pretty quickly. But maybe in case of APIs the codebase affected is only a tiny fraction of the total, so maybe db abstraction is not required everywhere, just in the API portion of the codebase.

Also, I'm not sure if this proposed solution should work on the cli level (i.e. you need to subprocess `sentry run api` or something) or the library level (i.e. you'd run an API server as a fixture within your test suite). I guess the former is more impactful, because that would allow for testing across languages, the latter is a nice to have for Python developers.

Last but not least - maybe this was proposed earlier - in that case I'm sorry for a duplicate issue, I tried searching for quite a while, but couldn't find anything.

**Comment by getsentry-release** [comment]
Routing to @getsentry/open-source for [triage](https://develop.sentry.dev/processing-tickets/#3-triage). ⏲️

**Comment by chadwhitacre** [comment]
@kokes Two thoughts:

1. Do you know about [self-hosted Sentry](https://develop.sentry.dev/self-hosted/)? This is our packaging of Sentry for small deploys like you're talking about. You could maintain a long-running instance for use in testing, or sort out dynamic creation of Sentry instances based on this. It's still heavy (it has all of the components necessary for full Sentry) but maybe it's easier to work with for your use case than devservices?
1. Maybe you could run [Relay in static mode](https://docs.sentry.io/product/relay/modes/#static-mode)? That's the ingest point in Sentry's [architecture](https://develop.sentry.dev/architecture/), I'm actually not sure how much validation Relay performs and I'm not sure how much validation you _want_ to be performed as part of your testing. But if you want to minimize the Sentry footprint in your testing then this might get you what you need. Fire off events to Relay and inspect kafka/redis/logs for expected output, or possibly just responses from Relay.

(P.S. Sorry for the late reply, I was [OOO](https://github.com/getsentry/self-hosted/issues/1390) last week.)

**Comment by kokes** [comment]
@chadwhitacre 

Thanks for the suggestions

1. We actually ran self hosted Sentry for quite a while, but as you note - it shares the same heavy footprint as devservices, so it's not really a viable alternative. One more tidbit to add - some of the dependencies are fairly outdated (e.g. Postgres 9.6), which triggers all sorts of security checks. But that's not the main point, just one more observation.
2. That sounds good. The intention here is to verify that a given payload would be ingested in Sentry - that is, if it returns a 200/204 and/or logs something someplace, it should be a perfectly viable option. Though as you say, there are different checks performed at various stages, so I'll have to investigate this. E.g. payload size limits seem to be in the API, not relay etc.

(We've since migrated from self hosted to Sentry.io and run our integration tests against it. Which... works, but I'm not a huge fan of firing requests against live instances in integration tests.)
